src/Runner.Worker/ActionRunner.cs

```diff
@@ -179,17 +179,15 @@ public async Task RunAsync()
             ExecutionContext.Debug("Loading env");
             var environment = new Dictionary<String, String>(VarUtil.EnvironmentVariableKeyComparer);
 
-            // Apply environment set using ##[set-env] first since these are job level env
-            foreach (var env in ExecutionContext.EnvironmentVariables)
+#if OS_WINDOWS
+            var envContext = ExecutionContext.ExpressionValues["env"] as DictionaryContextData;
+#else
+            var envContext = ExecutionContext.ExpressionValues["env"] as CaseSensitiveDictionaryContextData;
+#endif
+            // Apply environment from env context, env context contains job level env and action's evn block
+            foreach (var env in envContext)
```

Comment:
I moved the `step.env` to when we start the step, so I can include the `step.env` as part for the `env` context before evaluate displayname, if and step.inputs.

Reply:
Doesn't display name get calculated at the beginning of the job.

When calculating display name at beginning of a job, should check if script contains `env` and delay. We have logic on server to check for referenced contexts... Might be the key to solving hashFiles issue too.

Reply:
Should we evaluate `if` before `env`? Or is it because we need `env` to calculate display name, even when skipped?

Reply:
we re-evaluate displayname in `stepsrunner` before each step run: `actionRunner.TryEvaluateDisplayName(step.ExecutionContext.ExpressionValues, step.ExecutionContext)`, we need `env` to evaluate `if`, we don't calculate display name on skip.